Feat/allow different temperatures ts integrate

thomasloux opened the pull request:
## Summary

Change ts.integrate to accept temperature of size (n_system,), so that its system can have its own temperature.

## Checklist

Before a pull request can be merged, the following items must be checked:

* [x] Doc strings have been added in the [Google docstring format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google).
* [X] Run [ruff](https://beta.ruff.rs/docs/rules/#pydocstyle-d) on your code.
* [X] Tests have been added for any new functionality or bug fixes.

<!-- We highly recommended installing the `prek` hooks running in CI locally to speedup the development process. Simply run `pip install prek && prek install` to install the hooks which will check your code before each commit. -->

orionarcher commented on torch_sim/runners.py:
```diff
@@ -100,6 +100,67 @@ def _configure_batches_iterator(
+    # different temperatures per system, not per step.
+    if temps.shape[0] == initial_state.n_systems:
+        # Interpret as single-step multi-system temperatures → broadcast over steps
+        return temps.unsqueeze(0).expand(n_steps, -1)  # (n_steps, n_systems)
```
I honestly might just throw a well-documented error here and demand the user provide a 2D tensor if `n_systems == n_steps`. It's a rare edge case and the consequences of incorrectly guessing the default is potentially many hours of wasted debugging.

thomasloux replied:
I added a warning, let me know if you want to be more strict and raise an error

orionarcher reviewed (commented):
Awesome! This is a nice feature. I'd suggest adding a bit more testing fo the different temperature shapes though.

thomasloux commented:
Note for me: I need to add a test to check that it works with the autobatcher

thomasloux commented:
@orionarcher I modified the autobatcher.bin_index which was a list[dict[int, float]] which I think is just the results of the original use of the binpacking from the first version of autobatcher #39. I changed the bin_index to actually be list[list[indices]]. Otherwise
`for state, indices in batcher`: indices would be a super weird like {1: float, 2: float} and so you can't use it directly like: myTensor[indices]

orionarcher commented on torch_sim/runners.py:
```diff
@@ -123,7 +189,11 @@ def integrate[T: SimState](  # noqa: C901
+            Float: used for all steps and systems
+            1D array of length n_steps: used for each step
+            1D array of length n_systems: used for each system
+            2D array of shape (n_steps, n_systems): used for each step and system.
```
Could we add a test for the 2D array too for completeness?

orionarcher reviewed (approved):
Would be nice to have a test with a 2D array but otherwise LGTM
